### PR TITLE
194477 - create s100 ess api endpoint for updatessince

### PR DESCRIPTION
**Pull request by MastekRohitChettri**
This PBI consists changes of [PBI : 194477](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/194477)

- Added code changes
- Added unit test cases

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Validation/V2/UpdatesSinceValidator.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        private readonly IConfiguration _configuration;
+        private DateTime _sinceDateTime;
+
+        public UpdatesSinceValidator(IConfiguration configuration)
+        {
+            _configuration = configuration;
+
+            RuleFor(x => x.SinceDateTime)
+                .Must(x => x.IsValidRfc1123Format(out _sinceDateTime))
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Validation/V2/UpdatesSinceValidator.cs:27]: Valid date is in "yyyy-MM-ddTHH:mm:ss.fffZ" format (ISO 8601) please change wherever applicable.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+using UKHO.ExchangeSetService.API.Controllers;
+using UKHO.ExchangeSetService.API.Services;
+using UKHO.ExchangeSetService.Common.Models;
+using UKHO.ExchangeSetService.Common.Models.Response;
+using UKHO.ExchangeSetService.Common.Models.V2.Request;
+
+namespace UKHO.ExchangeSetService.API.UnitTests.Controllers
+{
+    [TestFixture]
+    public class ExchangeSetControllerTests
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs:20]: Can we add Unit test for passing null productIdentifier as well please.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Validation/V2/UpdatesSinceValidator.cs (hunk 1 above)
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Validation/V2/UpdatesSinceValidator.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs (hunk 2 above)
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs:24]: Added

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs [hunk 3]
```diff
@@ -852,9 +852,19 @@ public enum EventIds
...
+        /// </summary>
+        CreateUpdateSinceRequestStarted = 805212,
+        /// <summary>
+        /// 805213 - Request for creating update since request is completed.
+        /// </summary>
+        CreateUpdateSinceRequestCompleted = 805213,
+        /// <summary>
+        /// 805214 - Exception occurred for creating update since request.
+        /// </summary>
+        CreateUpdateSinceRequestException = 805214,
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs:867]: Can we rename these events to be more specific please.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/V2/Response/ExchangeSetStandardResponse.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        public Links Links { get; set; }
+
+        [JsonProperty("exchangeSetUrlExpiryDateTime")]
+        public DateTime? ExchangeSetUrlExpiryDateTime { get; set; }
+
+        [JsonProperty("requestedProductCount")]
+        public int RequestedProductCount { get; set; }
+
+        [JsonProperty("exchangeSetCellCount")]
+        public int ExchangeSetCellCount { get; set; }
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/V2/Response/ExchangeSetStandardResponse.cs:23]:
```suggestion
        [JsonProperty("exchangeSetProductCount")]
        public int ExchangeSetProductCount { get; set; }
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs (hunk 3 above)
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs]: Updated

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/V2/Response/ExchangeSetStandardResponse.cs (hunk 4 above)
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/V2/Response/ExchangeSetStandardResponse.cs]: Resolved

**Review by shirinbt14926 — CHANGES_REQUESTED** [review]
Minor comments please

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+            Action nullLogger = () => new ExchangeSetController(_fakeHttpContextAccessor, null, _fakeExchangeSetStandardService);
+            nullLogger.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("logger");
+
+            Action nullExchangeSetService = () => new ExchangeSetController(_fakeHttpContextAccessor, _fakeLogger, null);
+            nullExchangeSetService.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("exchangeSetService");
+        }
+
+        [Test]
+        public async Task WhenValidSinceDateTimeRequested_ThenPostUpdatesSince_Returns202Accepted()
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "UpdatesSince endpoint request for X-Correlation-ID : {correlationId} and ExchangeSetStandard : {exchangeSetStandard}").MustHaveHappened();
+
+            A.CallTo(_fakeLogger).Where(call => call.Method.Name == "Log"
+            && call.GetArgument<LogLevel>(0) == LogLevel.Information
+            && call.GetArgument<EventId>(1) == EventIds.PostUpdatesSinceRequestCompleted.ToEventId()
+            && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "UpdatesSince endpoint request for X-Correlation-ID : {correlationId} and ExchangeSetStandard : {exchangeSetStandard} Elapsed {Elapsed}").MustHaveHappened();
+        }
+
+        [Test]
+        public async Task WhenInValidSinceDateTimeRequested_ThenPostUpdatesSince_ReturnsBadRequest()
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ExchangeSetStandardServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            Action nullLogger = () => new ExchangeSetStandardService(null, _fakeUpdatesSinceValidator);
+            nullLogger.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("logger");
+
+            Action nullUpdatesSinceValidator = () => new ExchangeSetStandardService(_fakeLogger, null);
+            nullUpdatesSinceValidator.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("updatesSinceValidator");
+        }
+
+        [Test]
+        public async Task WhenValidSinceDateTimeRequested_ThenCreateUpdatesSince_Returns202Accepted()
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ExchangeSetStandardServiceTests.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+            A.CallTo(() => _fakeExchangeSetStandardService.CreateUpdatesSince(A<UpdatesSinceRequest>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored, A<CancellationToken>.Ignored))
+                .Returns(ServiceResponseResult<ExchangeSetStandardServiceResponse>.BadRequest(new ErrorDescription
+                {
+                    CorrelationId = Guid.NewGuid().ToString(),
+                    Errors = [new() { Source = "SinceDateTime", Description = "Provided sinceDateTime is either invalid or invalid format, the valid format is 'ISO 8601 format' (e.g. '2024-12-20T11:51:00.000Z')." },
+                                  new() { Source = "ProductIdentifier", Description = "ProductIdentifier must be valid value" },
+                                  new() { Source = "CallbackUri", Description = "Invalid callbackUri format." }],
+
+
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs:154]: Please remove blanks.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs (hunk 8 above)
Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ExchangeSetControllerTests.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs [hunk 9]
```diff
@@ -856,9 +856,32 @@ public enum EventIds
...
+        /// <summary>
+        /// 805212 - Request for updates since started.
+        /// </summary>
+        PostUpdatesSinceRequestStarted = 805212,
+        /// <summary>
+        /// 805213 - Request for updates since completed.
+        /// </summary>
+        PostUpdatesSinceRequestCompleted = 805213,
+        /// <summary>
+        /// 805214 - Creation of update since is started.
```
Comment (shirinbt14926) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs:872]:
```suggestion
        /// 805214 - Creation of updatesSince endpoint started.
```
Similarly make changes in other eventId description as well

Comment (MastekRohitChettri) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Logging/EventIds.cs]: Resolved

**Review by shirinbt14926 — APPROVED** [review]
LGTM.

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Review by harshal11869 — APPROVED** [review]
Looks Good.
